src/aero_kernel/src/syscall/ipc.rs

```diff
@@ -0,0 +1,161 @@
+ * along with Aero. If not, see <https://www.gnu.org/licenses/>.
+ */
+use crate::{
+    userland::scheduler::get_scheduler,
+    utils::{
+        sync::{BlockQueue, Mutex},
+        validate_slice_mut,
+    },
+};
+use aero_syscall::AeroSyscallError;
+use alloc::{collections::VecDeque, vec::Vec};
+use hashbrown::HashMap;
+use lazy_static::lazy_static;
+
+struct Message {
+    from: usize,
+    data: Vec<u8>,
+}
+struct MessageQueue {
+    pub queue: VecDeque<Message>,
+}
+
+lazy_static! {
+    static ref BLOCK_QUEUE: BlockQueue = BlockQueue::new();
+    static ref MESSAGES: Mutex<HashMap<usize, MessageQueue>> = Mutex::new(HashMap::new());
+}
+
+fn messagequeue_do_recieve(
+    pidptr: usize,
+    messageptr: usize,
+    messagesiz: usize,
+    msg: Message,
+) -> Result<usize, AeroSyscallError> {
+    let output =
+        validate_slice_mut(messageptr as *mut u8, messagesiz).ok_or(AeroSyscallError::EINVAL)?;
+
+    output
+        .split_at_mut(msg.data.len())
+        .0
+        .copy_from_slice(&msg.data);
+
+    unsafe {
+        (pidptr as *mut usize).write(msg.from);
+    }
+    Ok(msg.data.len())
+}
+pub fn send(pid: usize, message: usize, messagesiz: usize) -> Result<usize, AeroSyscallError> {
+    let payload =
+        validate_slice_mut(message as *mut u8, messagesiz).ok_or(AeroSyscallError::EINVAL)?;
+
+    let mut messagequeue = MESSAGES.lock();
+    match messagequeue.get_mut(&pid) {
+        Some(mq) => {
+            mq.queue.push_back(Message {
+                from: get_scheduler().current_task().pid().as_usize(),
+                data: payload.to_vec(),
+            });
+        }
+        None => {
+            messagequeue.insert(
+                pid,
+                MessageQueue {
+                    queue: [Message {
+                        from: get_scheduler().current_task().pid().as_usize(),
+                        data: payload.to_vec(),
+                    }]
+                    .into(),
+                },
+            );
+        }
+    }
+    // let bqueueitem = bqueue.get_mut(&pid).unwrap();
+    // bqueueitem.replace(Message {
+    //     from: get_scheduler().current_task().pid().as_usize(),
+    //     data: payload.to_vec(),
+    // });
+    BLOCK_QUEUE.notify_complete();
+    Ok(0)
+}
+pub fn recv(
+    pidptr: usize,
+    messageptr: usize,
+    messagemax: usize,
+    block: usize,
+) -> Result<usize, AeroSyscallError> {
+    let pid = get_scheduler().current_task().pid().as_usize();
+
+    if block == 0 {
+        // nonblocking read
+        let mut msgqueue = MESSAGES.lock();
+        match msgqueue.get_mut(&pid) {
+            Some(m) => {
+                let item = m
+                    .queue
+                    .pop_front()
+                    .expect("empty message queues should always be deleted!");
+                if item.data.len() > messagemax {
+                    m.queue.push_front(item);
+                    return Err(AeroSyscallError::E2BIG);
```

Comment:
Last expression in a block will be the return value so you can get rid of the return and trailing semi

Reply:
okay